packages/compiler/src/server/serverlib.ts

```diff
@@ -347,6 +367,91 @@ export function createServer(host: ServerHost): Server {
     }
   }
 
+  async function renameFiles(params: RenameFilesParams): Promise<void> {
+    const firstFilePath = params.files[0];
+    if (!firstFilePath) {
+      return;
+    }
+
+    // Update cache for renamed folders.
+    for (const file of params.files) {
+      const oldFilePath = await fileService.getPath({ uri: file.oldUri });
+      const newFilePath = await fileService.getPath({ uri: file.newUri });
+      const isDirRename = !file.oldUri.endsWith(".tsp");
+
+      if (isDirRename) {
+        const files = await listAllFilesInDir(compilerHost, newFilePath);
+        for (const file of files) {
+          const oldFile = resolvePath(oldFilePath, file);
```

Comment:
i think we can rebuild an array and push to the array when the file is .tsp file. and then we just use the new array below so that we only need to check files in it, no longer to worry about the directories which can help to make the following logic more clear.

Reply:
I think this hits the timing issue problem(which the cache refresh will also hit but not the end of the world right now for that).
As the actual renaming is happening in parallel with large rename of directories it could I think not collect the accurate list of files.

And as the comment above I think the current approach is better with startWith anyway

Reply:
we might want instead the cache to be able to invalidate a directory so we don't have to crawl those files but this just feels like a followup